teapots/classic-teapot/src/main/cpp/CMakeLists.txt

```diff
@@ -16,37 +16,32 @@
 
 cmake_minimum_required(VERSION 3.4.1)
 
-# build native_app_glue as a static lib
-add_library(native_app_glue STATIC
-    ${ANDROID_NDK}/sources/android/native_app_glue/android_native_app_glue.c)
-
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11 -Wall -fno-exceptions -fno-rtti")
-
-# build the ndk-helper library
-set(ndk_helper_dir ../../../../common/ndk_helper)
-add_subdirectory(${ndk_helper_dir} ndk_helper)
-
-# Export ANativeActivity_onCreate(), 
-# Refer to: https://github.com/android-ndk/ndk/issues/381.
-set(CMAKE_SHARED_LINKER_FLAGS
-    "${CMAKE_SHARED_LINKER_FLAGS} -u ANativeActivity_onCreate")
+# build the ndk-helper library into a common directory
```

Comment:
This looks like a hack to me, and I'm not sure we should be encouraging it. It violates layering by encoding knowledge about other projects here, even though this should be a standalone project as far as cmake is concerned. If there are things that are to be shared between multiple modules/projects, then I think that should be handled at a higher level (either a separate module/project in gradle, or something else).

Reply:
the new version is changed, thanks!

Reply:
I agree Android Studio shouldn't be involved in the factoring since it is a native-to-native dependency. I don't think @ggfan has any options that are better than add_directory. And, assuming there's a reason ndk_helper is shared as source rather than as .a or .so, I don't know of a better way than add_directory here even with cooperation from the NDK. @DanAlbert jic

Reply:
Surely there's a middle ground between this and an AAR. Is there no way to share components between multiple cmake projects in gradle?

Reply:
Yeah CDep could solve the problem: because it has header in the interface, aar is not really a solution for this.

I would like to keep this code change to be CMake centric: not adding another component into the mix. Like to prove that normal CMake should be able to get the work done ( CMake should be enough there is nothing magic here at all ).

Reply:
Just for completeness sake (recommend against this as it'll confuse lots of people):
https://cmake.org/cmake/help/v3.2/module/CPack.html
CPack technically would allow one to author a package using just cmake style stuff specifying the "install" location.  Then a consuming level app built separately with gradle module, can consume it by find_package() syntax.
Thus there does exist a solution that would be very similar to a cdep style solution.  However, getting it right is tricky to the point of "I personally haven't gotten it to work right."  I do use this style trick with the PS4 variant so I know it works, but cmake add_directory is much cleaner.

Reply:
cdep is *an* answer here, but that's really meant as a distribution channel. I wouldn't expect people to use it in a case like this where the dependency is expected to be built as a part of the project (though if ndk-helper were split out into its own project and a cdep package were maintained that was imported here, then I agree).

I was expecting something more along the lines of there being a CMake project for ndk-helper that gets built by gradle into its own project and then there's a FindNdkHelper.cmake that sets up the library and include paths. aiui that's the preferred way of importing externally built libraries into CMake.

Reply:
I actually am working on using cdep for exactly this kind of purpose (gradle plugin for authoring packages etc).  But that's a future thing.
What you're referencing is actually possible; gradle build that invokes cmake project for ndk-helper into its' own project.  What it'd take isn't what I'd call normal procedures, but here's a quick stab off the top of my head:
from module level gradle:
applicationVariants.all { variant ->
       variant.mergeAssets.doLast {
                copy { 
                   from: <there's a clever closure here involving variant and output files I don't know off top of my head>
                   to: "src/main/jniLibs"
                   include: *.so *.a
               }
            }
        }
 
Then either create a file called FindNdkHelper.cmake that looks like this:
set (LIBNDKHELPER_FOUND true)
set (LIBNDKHELPER_LIBRARIES ${CMAKE_CURRENT_BINARY_DIR}/ndkHelper)
set (LIBNDKHELPER_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/src/include)

Or one could set CMAKE_LIBRARY_PATH to be the output folder where ndkHelper gets built sending that to cmake either through the command line or through a set() command:
https://gitlab.kitware.com/cmake/community/wikis/doc/cmake/Useful-Variables

One could also have the original gradle invocation issue an "install" command to cmake's build process (either through a gradle custom command or through the gradle android studio plugin).
That would be the path of using cpack.

Here's an example for how I solved a somewhat similar problem (have gradle get cmake to output an xcode framework): https://github.com/gpx1000/ARKitARCoreBridge

I'm going to stay out of policy discussions, just here to offer help ;-).  I might suggest however, that if the flags in this case aren't necessary for the NdkHelper object, then they probably do belong in the CMAKE_CXX_FLAGS at the individual project level.  Let me know if I can be of more help (not wanting to get into any debates).

Reply:
Yeah, something along those lines is more what I would expect, but I was hoping there might be some out of the box support for doing it from gradle. Might be best to back out this part of the change (it's not really related to the rest of the commit anyway) and explore this further until we find the right solution.

Reply:
Maybe something like this? https://gitlab.kitware.com/cmake/community/wikis/doc/tutorials/Exporting-and-Importing-Targets

Reply:
this is built it into a third party lib, and then import it as an external lib. Which essentially is what cdep is ( and CDep goes one step further: release the binary, not source so no need to build at all)

Reply:
@DanAlbert That is an awesome idea.  However, I'm noticing that there's a requirement there of an "install" phase for CMake.  FWIW as far as I know, Android Studio's gradle plugin doesn't utilize the cmake install phase; only build.

However, the interesting thing about this as an idea is if, in the future, there were to be a cmake install phase available, then we can use CPack or any form of exporting/importing targets or even targeting non-android builds.  It would eliminate many of the requirements from other systems.  It also would mean that we could craft a CPack CDep package generator and eliminate the need for a separate gradle publishing plugin.